test/metabase/lib/datascript_test.cljc

```diff
@@ -0,0 +1,1100 @@
+          the-join (-> query :mbql.query/stage :mbql.stage/join first)]
+      (testing "has :mbql.join/source to the Products table"
+        (is (=? {:metadata.table/id (meta/id :products)}
+                (:mbql.join/source the-join))))
+      (testing "has :mbql.join/conditions included"
+        (is (=? [{:mbql.clause/operator :=
+                  :mbql/series          0
+                  :mbql.clause/argument
+                  [{:mbql/series 0, :mbql/ref {:metadata.field/id      (meta/id :orders :product-id)}}
+                   {:mbql/series 1, :mbql/ref {:metadata.column/source the-join
+                                               :metadata.column/name   "ID"
+                                               :metadata.column/mirror {:metadata.field/id (meta/id :products :id)}}}]}]
+                (:mbql.join/condition the-join))))
+      (testing "has :mbql/series 0"
+        (is (=? 0 (:mbql/series the-join))))
+
+      (testing "output columns"
+        (testing "work directly"
+          (is (=? (expected-join-fields the-join :products)
+                  (ld/returned-columns-method the-join))))
+        (testing "are included in the stage's"
+          (testing "returned-columns"
+            (is (=? (concat (expected-fields :orders)
+                            (expected-join-fields the-join :products))
+                    (ld/returned-columns query -1))))
+          (testing "visible-column-groups"
+            (is (=? [{:lib/type                              :metadata/column-group
+                      :metabase.lib.column-group/group-type  :group-type/main
+                      :metabase.lib.column-group/columns     (expected-fields :orders)}
+                     {:lib/type                              :metadata/column-group
+                      :metabase.lib.column-group/group-type  :group-type/join.explicit
+                      :metabase.lib.column-group/columns     (expected-join-fields the-join :products)}
+                     {:lib/type                              :metadata/column-group
+                      :metabase.lib.column-group/group-type  :group-type/join.implicit
+                      :metabase.lib.column-group/foreign-key {:metadata.field/id (meta/id :orders :user-id)}
+                      :metabase.lib.column-group/columns     sequential?}
+                     {:lib/type                              :metadata/column-group
+                      :metabase.lib.column-group/group-type  :group-type/join.implicit
+                      :metabase.lib.column-group/foreign-key {:metadata.field/id (meta/id :orders :product-id)}
+                      :metabase.lib.column-group/columns     sequential?}]
+                    (ld/visible-column-groups query -1)))))))))
+
+(deftest ^:parallel explicit-joins-test-forked-joins
+  (testing "central table joined to two others"
+    (let [;; Orders joined to Products on PRODUCT_ID, and to Reviews on Orders.USER_ID
```

Comment:
Bunch of tests for join contortions here and below - various cases of duplicate joins.